Nama  : Tiur Juliana Sitorus
NIM   : 42324048
Prodi : DIII Teknologi Informasi 2024
Mata Kuliah : Kecerdasan Buatan

Tugas
Prediksi apakah atlet dengan data berikut masuk draft yes/no
1.	input = [7.5, 7.0]
2.	input = [4.0, 6.5]

Langkah 1: Menyiapkan data

In [1]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

df = pd.read_csv('athletes.xls')

X = df[['speed', 'agility']].values
en = LabelEncoder()
y = en.fit_transform(df['draft'].values)

X_train, X_test = train_test_split(X, test_size=0.2, random_state=0)
y_train, y_test = train_test_split(y, test_size=0.2, random_state=0)

Penjelasan: Bagian awal kode memanggil library yang dibutuhkan. Dataset dibaca dengan read_csv, karena walaupun ekstensinya .xls, isi file-nya berformat CSV. Kolom speed dan agility dijadikan fitur X, sedangkan kolom draft dijadikan target y. LabelEncoder mengubah "no" menjadi 0 dan "yes" menjadi 1. Terakhir, data dibagi menjadi 16 data latih dan 4 data uji. random_state=0 membuat hasil pembagiannya selalu sama setiap kali kode dijalankan.

Langkah 2: Mencari nilai K terbaik

In [2]:
for k in range(1, 11):
    model = KNeighborsClassifier(n_neighbors=k, metric='euclidean')
    model.fit(X_train, y_train)
    akurasi = accuracy_score(y_test, model.predict(X_test))
    print(f"K={k}  akurasi={akurasi:.2f}")

K=1  akurasi=0.75
K=2  akurasi=0.75
K=3  akurasi=1.00
K=4  akurasi=1.00
K=5  akurasi=0.75
K=6  akurasi=0.75
K=7  akurasi=0.75
K=8  akurasi=0.75
K=9  akurasi=0.75
K=10  akurasi=0.50


Penjelasan: Perulangan ini mencoba nilai K dari 1 sampai 10. Untuk setiap nilai K, model dilatih lalu diuji, kemudian akurasinya ditampilkan. Hasilnya, akurasi tertinggi (1.00) diperoleh pada K = 3 dan K = 4. Kita memilih K = 3 karena merupakan bilangan ganjil. Dengan dua kelas (yes/no), jumlah tetangga yang ganjil membuat voting tidak mungkin seri.

Langkah 3: Memprediksi atlet baru

In [3]:
knn_atlet = KNeighborsClassifier(n_neighbors=3, metric='euclidean')
knn_atlet.fit(X_train, y_train)

inputs_atlet = [[7.5, 7.0], [4.0, 6.5]]
for inp in inputs_atlet:
    pred = knn_atlet.predict([inp])
    jarak, idx = knn_atlet.kneighbors([inp])
    print(f"Input {inp} -> draft: {en.inverse_transform(pred)[0]}")
    for d, i in zip(jarak[0], idx[0]):
        label = en.inverse_transform([y_train[i]])[0]
        print(f"   tetangga {X_train[i]}  label={label}  jarak={d:.3f}")

Input [7.5, 7.0] -> draft: yes
   tetangga [8.25 8.5 ]  label=no  jarak=1.677
   tetangga [5.5  6.75]  label=yes  jarak=2.016
   tetangga [5.75 8.75]  label=yes  jarak=2.475
Input [4.0, 6.5] -> draft: yes
   tetangga [4.75 6.25]  label=yes  jarak=0.791
   tetangga [3.5  5.25]  label=no  jarak=1.346
   tetangga [5.5  6.75]  label=yes  jarak=1.521


Penjelasan: Model KNN dibuat dengan 3 tetangga dan jarak Euclidean, lalu dilatih dengan fit(). Dua data atlet baru disimpan di inputs_atlet, kemudian diprediksi satu per satu. Fungsi predict() menghasilkan angka 0 atau 1, dan inverse_transform() mengubahnya kembali menjadi "no" atau "yes". Fungsi kneighbors() menampilkan tiga atlet terdekat beserta label dan jaraknya, sehingga kita bisa melihat alasan model memilih jawaban tersebut.

Output:
Input [7.5, 7.0] -> draft: yes
   tetangga [8.25 8.5 ]  label=no  jarak=1.677
   tetangga [5.5  6.75]  label=yes  jarak=2.016
   tetangga [5.75 8.75]  label=yes  jarak=2.475
Input [4.0, 6.5] -> draft: yes
   tetangga [4.75 6.25]  label=yes  jarak=0.791
   tetangga [3.5  5.25]  label=no  jarak=1.346
   tetangga [5.5  6.75]  label=yes  jarak=1.521

Hasil dan analisis :

Atlet pertama [7.5, 7.0] diprediksi masuk draft (yes). Tetangga paling dekatnya sebenarnya berlabel no, yaitu atlet [8.25, 8.5], tetapi dua tetangga lainnya berlabel yes. Karena voting-nya 2 lawan 1, hasilnya yes. Secara logika ini masuk akal, karena speed dan agility atlet ini sama-sama tinggi.

Atlet kedua [4.0, 6.5] juga diprediksi masuk draft (yes) dengan voting 2 lawan 1. Tetangga terdekatnya [4.75, 6.25] berlabel yes dan jaraknya sangat dekat (0.791). Meski begitu, posisi atlet ini berada di perbatasan antara kelompok yes dan no. Kalau K diubah menjadi 5, dua tetangga tambahannya berlabel no, sehingga hasilnya berubah menjadi no. Ini menunjukkan bahwa pemilihan nilai K bisa memengaruhi hasil prediksi, terutama untuk data yang posisinya di perbatasan antar kelas.

Untuk dataset iris, prediksi input data berikut ini masuk ke species apa:
1.	input_data = [[7.1, 1.2, 5.0, 2.0]]
2.	input_data = [[5.1, 3.5, 1.4, 0.2]]
3.	input_data = [[6.4, 3.7, 4.2, 2.2]]

Langkah 1: Menyiapkan data

In [4]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

iris = load_iris()
df_iris = pd.DataFrame(iris.data, columns=iris.feature_names)

lab = LabelEncoder()
y_iris = lab.fit_transform(iris.target_names[iris.target])

Xi_train, Xi_test = train_test_split(df_iris, test_size=0.2, random_state=0)
yi_train, yi_test = train_test_split(y_iris, test_size=0.2, random_state=0)

Penjelasan: Bagian awal kode memanggil library yang dibutuhkan. Dataset Iris dimuat dengan load_iris(), lalu diubah menjadi DataFrame agar setiap kolom punya nama fitur. Nama spesies diubah menjadi angka dengan LabelEncoder: setosa menjadi 0, versicolor menjadi 1, dan virginica menjadi 2. Setelah itu data dibagi menjadi 120 data latih dan 30 data uji. random_state=0 membuat hasil pembagiannya selalu sama setiap kali kode dijalankan.

Langkah 2: Melatih model dan mengecek akurasi

In [5]:
knn_iris = KNeighborsClassifier(n_neighbors=3, metric='euclidean')
knn_iris.fit(Xi_train, yi_train)

akurasi = accuracy_score(yi_test, knn_iris.predict(Xi_test))
print("Akurasi data uji:", akurasi)

Akurasi data uji: 0.9666666666666667


Penjelasan: Model KNN dibuat dengan K = 3 dan jarak Euclidean, sama seperti di modul, lalu dilatih dengan fit(). Sebelum dipakai untuk data baru, model diuji dulu dengan data uji. Akurasinya 0.967, artinya 29 dari 30 bunga ditebak dengan benar, jadi model sudah cukup bagus untuk memprediksi data baru.

Langkah 3: Memprediksi bunga baru

In [6]:
inputs_iris = [[7.1, 1.2, 5.0, 2.0],
               [5.1, 3.5, 1.4, 0.2],
               [6.4, 3.7, 4.2, 2.2]]

for inp in inputs_iris:
    data = pd.DataFrame([inp], columns=iris.feature_names)
    pred = knn_iris.predict(data)
    jarak, idx = knn_iris.kneighbors(data)
    tetangga = lab.inverse_transform(yi_train[idx[0]])
    print(f"Input {inp} -> species: {lab.inverse_transform(pred)[0]}")
    print(f"   3 tetangga terdekat: {tetangga.tolist()}  jarak: {np.round(jarak[0], 3)}")

Input [7.1, 1.2, 5.0, 2.0] -> species: virginica
   3 tetangga terdekat: ['versicolor', 'virginica', 'virginica']  jarak: [1.52  1.53  1.568]
Input [5.1, 3.5, 1.4, 0.2] -> species: setosa
   3 tetangga terdekat: ['setosa', 'setosa', 'setosa']  jarak: [0.    0.1   0.141]
Input [6.4, 3.7, 4.2, 2.2] -> species: versicolor
   3 tetangga terdekat: ['versicolor', 'versicolor', 'versicolor']  jarak: [0.837 0.883 1.01 ]


Penjelasan: Tiga data bunga baru disimpan di inputs_iris. Setiap data dibungkus ke dalam DataFrame dengan nama kolom yang sama seperti data latih. Ini perlu karena model dilatih memakai DataFrame, sehingga tanpa nama kolom akan muncul warning. Fungsi predict() menghasilkan angka 0, 1, atau 2, lalu inverse_transform() mengubahnya kembali menjadi nama spesies. Fungsi kneighbors() menampilkan tiga bunga terdekat beserta jaraknya, sehingga kita bisa melihat alasan model memilih spesies tersebut.

Output:
Input [7.1, 1.2, 5.0, 2.0] -> species: virginica
   3 tetangga terdekat: ['versicolor', 'virginica', 'virginica']  jarak: [1.52  1.53  1.568]
Input [5.1, 3.5, 1.4, 0.2] -> species: setosa
   3 tetangga terdekat: ['setosa', 'setosa', 'setosa']  jarak: [0.    0.1   0.141]
Input [6.4, 3.7, 4.2, 2.2] -> species: versicolor
   3 tetangga terdekat: ['versicolor', 'versicolor', 'versicolor']  jarak: [0.837 0.883 1.01 ]

Hasil dan analisis

Bunga pertama [7.1, 1.2, 5.0, 2.0] diprediksi virginica. Dari tiga tetangga terdekatnya, dua virginica dan satu versicolor, sehingga voting-nya 2 lawan 1. Prediksi ini kurang meyakinkan. Nilai sepal width 1.2 jauh di bawah data Iris pada umumnya, yang minimumnya sekitar 2.0, dan jarak ke tetangganya juga cukup jauh (sekitar 1.5). Artinya bunga ini sebenarnya tidak terlalu mirip dengan bunga mana pun di dataset, tetapi KNN tetap memberi jawaban.

Bunga kedua [5.1, 3.5, 1.4, 0.2] diprediksi setosa. Prediksi ini sangat yakin, karena ketiga tetangganya setosa. Jarak ke tetangga pertama bahkan 0, artinya data ini sama persis dengan salah satu bunga di dataset.

Bunga ketiga [6.4, 3.7, 4.2, 2.2] diprediksi versicolor, karena ketiga tetangga terdekatnya versicolor. Nilai petal length 4.2 memang khas versicolor, yang berada di antara setosa (kecil) dan virginica (besar).

Secara keseluruhan, KNN menebak spesies bunga baru berdasarkan kemiripan ukurannya dengan bunga yang sudah ada. Semakin dekat jarak dan semakin kompak voting tetangganya, semakin bisa dipercaya hasil prediksinya.